### PR TITLE
Allows replacement of any reference with another one.

### LINKED ISSUE #240: add CtReference.getParent() and replace()
Why there are no methods like `CtReference. getParent()` and `replace()`. I can change all reference's fields, during scanning, but cannot change the reference class.

I stuck with implementing "generic specialization" properly, where I need to replace all CtTypeParameterReferences with CtTypeReferences.

### PR DESCRIPTION
**Pull request by GerardPaligot**
> **Note:** This PR introduces `CtObject` interface in the second commit and breaks compatibility with the usage of `getParent()` method.

Closes #240

### CODE REVIEW

**Comment by monperrus** [comment]
Thanks for this PR.

I disagree with the introduction of CtObject:

- the difference of responsibility between CtElement and CtObject is unclear, even in the names.
- the necessary changes in the tests with the casts to CtElement are not nice

What about a simple CtReplaceable interface for both elements and references?

**Comment by GerardPaligot** [comment]
> the difference of responsibility between CtElement and CtObject is unclear, even in the names.

I can replace it by `CtReplaceable`.

> the necessary changes in the tests with the casts to CtElement are not nice

I agree but necessary. Now, it is possible to have a `CtElement` or a `CtReference` as parent.

> What about a simple CtReplaceable interface for both elements and references?

It's already the case but named `CtObject`. I can change it.

**Comment by GerardPaligot** [comment]
If continuous integrations are ok, I'm okay too.

But if we merge this PR, I vote to make a major version at the next release.

**Comment by GerardPaligot** [comment]
The continuous integration server non regression at large scale is okay with this PR too.

capture d ecran 2015-11-19 a 11 25 30

File: src/test/java/spoon/test/replace/ReplaceTest.java [hunk 1]
```diff
@@ -216,4 +235,106 @@ public void testReplaceStatement() {
...
+				return reference.getActualTypeArguments().size() > 0 && super.matches(reference);
+			}
+		});
+
+		for (CtTypeReference reference : references) {
+			assertTrue(reference.getActualTypeArguments().size() > 0);
+		}
+
+		for (CtTypeReference<?> reference : references) {
+			reference.replace(factory.Type().createReference(reference.getQualifiedName()));
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/replace/ReplaceTest.java]: replace only one reference.

File: src/test/java/spoon/test/replace/ReplaceTest.java [hunk 2]
```diff
@@ -216,4 +235,106 @@ public void testReplaceStatement() {
...
+
+		aMethod.getType().replace(factory.Type().DOUBLE_PRIMITIVE);
+
+		assertEquals(factory.Type().DOUBLE_PRIMITIVE, aMethod.getType());
+	}
+
+	@Test
+	public void testReplaceAllTypeRefenceWithGenerics() throws Exception {
+		// contract: replace all type references with a generic to the same type reference without generics.
+		final List<CtTypeReference> references = Query.getReferences(factory, new ReferenceTypeFilter<CtTypeReference>(CtTypeReference.class) {
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/replace/ReplaceTest.java]: factory more explicit.

**Comment by GerardPaligot** [comment]
TODO: Change to 5.0.0-SNAPSHOT in pom.xml

**Comment by GerardPaligot** [comment]
TODO: Adds some test for parent of a CtReference.

File: src/test/java/spoon/test/replace/ReplaceTest.java [hunk 3]
```diff
@@ -216,4 +235,106 @@ public void testReplaceStatement() {
...
+	public void testReplaceAParameterReferenceToFieldReference() throws Exception {
+		// contract: replace a parameter reference to a field reference.
+		final CtType<Tacos> aTacos = factory.Type().get(Tacos.class);
+
+		final CtInvocation inv = aTacos.getMethodsByName("m3").get(0).getElements(new TypeFilter<>(CtInvocation.class)).get(0);
+		final CtVariableRead<?> variableRead = (CtVariableRead<?>) inv.getArguments().get(0);
+		final CtParameterReference<?> aParameterReference = (CtParameterReference<?>) variableRead.getVariable();
+		final CtFieldReference<?> aFieldReference = aTacos.getField("field").getReference();
+
+		assertEquals(variableRead.getVariable(), aParameterReference);
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/replace/ReplaceTest.java]: Flip expected, actual

**Comment by monperrus** [comment]
I agree with the jump to 5.0.0-SNAPSHOT, even if the tests show that nothing specified is broken, it may impact client processors of CtElement

**Comment by GerardPaligot** [comment]
TODO: Updates documentation

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Comment by GerardPaligot** [comment]
The continuous integration server non regression at large scale is always okay with this PR too.

capture d ecran 2015-12-02 a 10 08 06

**Pull request merged**